include/dlaf/eigensolver/backtransformation.h

```diff
@@ -0,0 +1,51 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2021, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <blas.hh>
+#include "dlaf/eigensolver/backtransformation/mc.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+
+namespace dlaf {
+namespace eigensolver {
+
+/// Eigenvalue back-transformation implementation on local memory.
+///
+/// It computes Q C, where Q = HH(1) HH(2) ... HH(m-mb)
+/// (HH(j) is the House-Holder transformation (I - v tau vH)
+/// defined by the j-th element of tau and the HH reflector stored in the j-th column of the matrix V.
+///
+/// @param mat_c contains the (m x n) matrix C (blocksize (mb x nb)), while on exit it contains Q C.
+/// @param mat_v is (m x m) matrix with blocksize (mb x mb), which contains the Householder reflectors.
+/// The j-th HH reflector is v_j = (1, V(mb + j : n, j)).
+/// @param taus is a (blocked) vector of size m (blocksize mb). The j-th element is the scaling factor
+/// for the j-th HH tranformation.
+/// @pre mat_c is not distributed,
+/// @pre mat_v is not distributed.
+template <Backend backend, Device device, class T>
+void backTransformation(Matrix<T, device>& mat_c, Matrix<const T, device>& mat_v,
+                        common::internal::vector<hpx::shared_future<common::internal::vector<T>>> taus) {
+  DLAF_ASSERT(matrix::local_matrix(mat_c), mat_c);
+  DLAF_ASSERT(matrix::local_matrix(mat_v), mat_v);
+  DLAF_ASSERT(mat_c.blockSize().rows() == mat_v.blockSize().rows(), mat_c, mat_v);
+
+  const SizeType m = mat_v.size().rows();
+  const SizeType mb = mat_v.blockSize().rows();
+  const auto nr_reflectors =
+      (mb == 1 || m % mb == 1) ? (mat_v.nrTiles().rows() - 2) : (mat_v.nrTiles().rows() - 1);
+  DLAF_ASSERT(taus.size() == nr_reflectors, taus.size(), mat_v, nr_reflectors);
```

Comment:
I'm sure there's a more elegant way to compute `nr_reflectors`. Suggestions welcome.

Reply:
I suggest to use `nr_reflector_blocks` as in the implementation. And it can be simply computed as `(m - mb - 1) / mb`.

Reply:
I've actually tried this yesterday as a first attempt, but it didn't work. 
This was the code:
`  const SizeType m = mat_v.size().rows();
  const SizeType mb = mat_v.blockSize().rows();
  DLAF_ASSERT(taus.size() == ((m - mb - 1)/mb), taus.size(), mat_v, ((m - mb - 1)/mb))`
Is that correct?
Because when I run it I got the following error:
`[ERROR] /home/lara/Documents/DLAF/DLA-Future_lara-eigenv-backtrans/include/dlaf/eigensolver/backtransformation.h:45 : void dlaf::eigensolver::backTransformation(dlaf::matrix::Matrix<T, device>&, dlaf::matrix::Matrix<const T, device>&, dlaf::common::internal::vector<hpx::lcos::shared_future<dlaf::common::internal::vector<T> > >) [with dlaf::Backend backend = (dlaf::Backend)0; dlaf::Device device = (dlaf::Device)0; T = float]
taus.size() == ((m - mb - 1)/mb)
1
size=(4, 4), block_size=(2, 2), tiles_grid=(2, 2), rank_index=(0, 0), comm_grid=(1, 1)
0

terminate called without an active exception
`
Can you please explain what I'm doing wrong?

Reply:
Sorry I didn't consider corner cases... a ceilDiv is needed here:
SizeType nr_reflector_blocks = std::max(0, ceildiv(m - mb - 1, mb));